### PR TITLE
Add static bfd monitor, plus tests

### PR DESCRIPTION
**Pull request by arp-est**
https://github.com/Nordix/Meridio-2/issues/104
https://github.com/Nordix/Meridio-2/issues/115

AI summary:
```  Static+BFD Readiness Gates

  What it does

  Extends the connectivity monitoring to support static routing with BFD. Previously, readiness gates (meridio-2.nordix.org/ipv4-connectivity / ipv6-connectivity) only worked for
  BGP sessions. Now, for GatewayRouters using static routing with BFD configured, the readiness gate reflects the BFD session health — if BFD goes down, the protocol is marked as
  not established, the gate goes False, and the ENC controller stops routing traffic through that LB pod.

  How it works

  1. Monitor queries both birdc show protocols and birdc show bfd sessions each tick
  2. applyBfdState() in run.go mutates static protocol statuses:
    - Static + BFD configured + BFD session Up → Info = "Established"
    - Static + BFD configured + BFD session Down/missing → State = Down
    - Static + no BFD → Info = "Established" (always connected)
    - BGP → untouched (BIRD already reports correct state)

  3. All downstream logic (protocolsUp, ClassifyConnectivityByFamily, gate manager) just uses IsEstablished() uniformly — no protocol-specific branching needed

  Files modified

  - internal/bird/monitor.go — BfdSession struct, parseBfdOutput(), BfdSessions on MonitorStatus, BFD query in checkStatus(), BgpInfoEstablished constant
  - internal/controller/router/connectivity.go — unchanged (original simple ClassifyConnectivityByFamily)
  - cmd/router/cmd/run.go — applyBfdState(), routers fetched earlier in monitor loop, passed to mutation before downstream logic

  Tests

  - cmd/router/cmd/run_test.go — unit tests for applyBfdState (8 cases: BFD up/down/missing, no BFD, BGP untouched, mixed protocols, etc.)
  - api/v1alpha1/gatewayrouter_validation_test.go — added test confirming empty static: {} (no BFD) is accepted by the API

  E2E suite

  - test/e2e/suites/separate-static-appnetwork/ — new suite with static+BFD routing, VIPs 110.0.0.1/110.0.0.2, VLANs 1100/1200
  - VPN gateway configured with 10 static routes per VIP with BFD — allows scaling to 10 replicas without gateway reconfiguration
  - NADs use range_start/range_end to constrain LB pod IPs to .1-.10
  - Added to SUITES_IPV4 in test/e2e/Makefile and testCases in e2e_suite_appnetwork_test.go
  - Updated all e2e routing.yaml files to explicitly set protocol: "BGP" or "Static"
  - Updated VPN gateway (init.sh, bird-gw.conf) addresses from 169.254.60/61 to 169.254.110/111

  API fixes

  - Added +kubebuilder:validation:Format="" + Maximum=4294967295 + Minimum=0 on RemoteASN/LocalASN to allow 4-byte ASNs
  - Removed the immutability of BGPSpec, StaticSpec, and Protocol fields.
  ```

### CODE REVIEW

**Review by bszirtes — COMMENTED** [review]
The e2e test part looks good.

File: cmd/router/cmd/run.go [hunk 1]
```diff
@@ -244,6 +244,17 @@ func monitorConnectivity(
...
 
+			// Fetch routers from informer cache (local memory read)
+			routers, err := getGatewayRoutersFromCache(ctx, mgr.GetClient(), cfg.GatewayName, cfg.GatewayNamespace)
+			if err != nil {
+				log.Error(err, "failed to get gateway routers from cache")
+				continue
+			}
+
+			// Mutate static protocol states based on BFD session status.
+			// After this, all protocols reflect true connectivity (State == Down if BFD is down).
```
Comment (ljkiraly) [inline cmd/router/cmd/run.go]: If I understand correctly the function sets the p.Info (not p.State) or which State is referred here?

File: internal/bird/monitor.go [hunk 2]
```diff
@@ -157,6 +174,49 @@ func hasConnectivity(protocols []ProtocolStatus) bool {
...
+			continue
+		}
+
+		fields := strings.Fields(line)
+		if len(fields) < 3 {
+			continue
+		}
+
+		// Skip protocol name lines like "bfd1:"
+		if len(fields) == 1 && strings.HasSuffix(fields[0], ":") {
```
Comment (ljkiraly) [inline internal/bird/monitor.go]: This filter is never used. The previous one (in line 201) catch the protocol name also. 
len (["bfd1:"]) < 3  == true (1 field)
Remove the dead code.

File: api/v1alpha1/gatewayrouter_types.go [hunk 3]
```diff
@@ -58,10 +58,16 @@ type GatewayRouterSpec struct {
 
 type BgpSpec struct {
 	// The ASN number of the Gateway Router
+	// +kubebuilder:validation:Format=""
+	// +kubebuilder:validation:Maximum=4294967295
```
Comment (ljkiraly) [inline api/v1alpha1/gatewayrouter_types.go:64]: Add some explanation to the comment. 
Validation is correct, allows full uint32 range for 4-byte ASNs.

File: api/v1alpha1/gatewayrouter_types.go [hunk 4]
```diff
@@ -58,10 +58,16 @@ type GatewayRouterSpec struct {
...
+	// +kubebuilder:validation:Minimum=0
 	// +required
 	RemoteASN uint32 `json:"remoteASN"`
 
 	// The ASN number of the system where the Attractor FrontEnds locates
+	// +kubebuilder:validation:Format=""
+	// +kubebuilder:validation:Maximum=4294967295
```
Comment (ljkiraly) [inline api/v1alpha1/gatewayrouter_types.go:76]: Add some explanation to the comment.
Validation is correct, allows full uint32 range for 4-byte ASNs.

File: cmd/router/cmd/run.go [hunk 5]
```diff
@@ -307,3 +313,45 @@ func protocolsUp(protocols []bird.ProtocolStatus) int {
...
+		gr := routerByName[p.Name]
+		if gr == nil {
+			continue
+		}
+
+		if gr.Spec.Static != nil && gr.Spec.Static.BFD != nil {
+			if bfdByIP[gr.Spec.Address] {
+				p.Info = bird.BgpInfoEstablished
+			} else {
+				p.Info = bird.BfdInfoDown
```
Comment (ljkiraly) [inline cmd/router/cmd/run.go:353]: A debug log when BFD causes a static protocol to be marked down helps production debugging.
```suggestion
				p.Info = bird.BfdInfoDown
				log.V(1).Info("BFD down for static protocol", "protocol", p.Name, "address", gr.Spec.Address)
```
Requires passing a `logr.Logger` into `applyBfdState` (add `log logr.Logger` parameter), or alternatively use `slog` package-level logging. 
Could be a follow-up.

**Review by ljkiraly — COMMENTED** [review]
Excellent design and functionally good.
After minor fixes can be approved. 
Beside my online suggestions one test coverage gap:
- Consider to add unit tests for `parseBfdOutput` - at minimum test standard BIRD output format, empty output, and IPv6 sessions. `parseBfdOutput` only tested indirectly through `applyBfdState`. 
Should test:
   - Standard BIRD BFD output with multiple sessions
   - Empty output (no BFD configured in BIRD)
   - Output with only headers (BFD protocol exists but no sessions)
   - IPv6 BFD sessions
   - Malformed lines

File: cmd/router/cmd/run.go (hunk 1 above)
Comment (arp-est) [reply cmd/router/cmd/run.go]: Ah, yes, originally the AI set state := Down, but later I've changed it to use the Info field instead, I didn't want it to mutate the state field itself.
I'll update the comment

File: cmd/router/cmd/run.go (hunk 5 above)
Comment (arp-est) [reply cmd/router/cmd/run.go:353]: I'll pass the log as parameter.

File: internal/bird/monitor.go (hunk 2 above)
Comment (arp-est) [reply internal/bird/monitor.go]: removed

File: api/v1alpha1/gatewayrouter_types.go (hunk 3 above)
Comment (arp-est) [reply api/v1alpha1/gatewayrouter_types.go:64]: done

File: api/v1alpha1/gatewayrouter_types.go (hunk 4 above)
Comment (arp-est) [reply api/v1alpha1/gatewayrouter_types.go:76]: done

File: cmd/router/cmd/run.go
Comment (zolug) [inline cmd/router/cmd/run.go]: HasConnectivity does not respect BFD status in case of Static, thus logging can be misleading. No need to pre-calculate it in checkStatus(), especially since count conveys the same meaning. Can be dropped from status alltogether.

File: cmd/router/cmd/run.go [hunk 6]
```diff
@@ -307,3 +314,47 @@ func protocolsUp(protocols []bird.ProtocolStatus) int {
...
+		if p.Proto != "Static" || !p.State.IsUp() {
+			continue
+		}
+
+		gr := routerByName[p.Name]
+		if gr == nil {
+			continue
+		}
+
+		if gr.Spec.Static != nil && gr.Spec.Static.BFD != nil {
```
Comment (zolug) [inline cmd/router/cmd/run.go:349]: Would be nice to match against the interface as well. In meridio v1 there could be only 1 external interface per router container, but here there could be many at least in theory.

File: internal/bird/monitor.go
Comment (rubasov) [inline internal/bird/monitor.go]: Is this TODO resolved by this PR? If yes, please remove it.

File: cmd/router/cmd/run.go (hunk 6 above)
Comment (arp-est) [reply cmd/router/cmd/run.go:349]: added the interface part

File: cmd/router/cmd/run.go
Comment (arp-est) [reply cmd/router/cmd/run.go]: removed hasConnectivity

**Comment by arp-est** [comment]
> Excellent design and functionally good. After minor fixes can be approved. Beside my online suggestions one test coverage gap:
> 
>     * Consider to add unit tests for `parseBfdOutput` - at minimum test standard BIRD output format, empty output, and IPv6 sessions. `parseBfdOutput` only tested indirectly through `applyBfdState`.
>       Should test:
>       
>       * Standard BIRD BFD output with multiple sessions
>       * Empty output (no BFD configured in BIRD)
>       * Output with only headers (BFD protocol exists but no sessions)
>       * IPv6 BFD sessions
>       * Malformed lines

added tests

File: internal/bird/monitor.go [hunk 7]
```diff
@@ -110,6 +118,15 @@ func (b *Bird) checkStatus(ctx context.Context) MonitorStatus {
 	}
 
 	status.Protocols = parseProtocolOutput(string(out))
+
+	bfdCmd := exec.CommandContext(ctx, "birdc", "-s", b.SocketPath, "show", "bfd", "sessions")
+	bfdOut, err := bfdCmd.CombinedOutput()
+	if err != nil {
+		log.V(1).Info("birdc bfd query failed", "error", err)
```
Comment (rubasov) [inline internal/bird/monitor.go:122]: It can be helpful in debugging if log lines like these always contain the command that failed, something like:

`Info("birdc failed", "cmd", bfdCmd, "error", err)`

Line 116 is not added by this PR, but this comment may apply there too.

**Comment by arp-est** [comment]
Added a follow up issue to discuss the future of the "Gateway established", "-//-lost" printouts
https://github.com/Nordix/Meridio-2/issues/223

File: internal/bird/monitor.go [hunk 8]
```diff
@@ -147,14 +160,42 @@ func parseProtocolOutput(output string) []ProtocolStatus {
...
+	State     string // "Up", "Down", "Init", etc.
+}
+
+// IsUp returns true if the BFD session state is "Up".
+func (b BfdSession) IsUp() bool {
+	return b.State == "Up"
+}
+
+// parseBfdOutput parses the output of `birdc show bfd sessions`.
+func parseBfdOutput(output string) []BfdSession {
```
Comment (rubasov) [inline internal/bird/monitor.go:175]: * This function could have been named parseShowBfdSessions.

* For maintainers coming back to this it can be helpful to add an example of the output we are parsing as a comment here.

* By the way I just found the '-v' flag for birdc in the bird docs:
https://bird.network.cz/doc/bird-4.html

Which is probably a feature to help parse the output of birdc. Here's an example I just made up, which leaves just a table to parse:

```
# birdcl -r -v -s /run/bird1/bird.ctl show protocols all | egrep '^1002-' | cut -d- -f2- 
kernel1    Kernel     master4    up     07:57:05.689  
static_routes Static     master4    up     07:57:05.689  
bird2      BGP        ---        up     08:28:11.415  Established   
device1    Device     ---        up     07:57:05.689  
direct1    Direct     ---        up     07:57:05.689  
```

**Review by rubasov — COMMENTED** [review]
As a new contributor I started reviewing this to learn about the codebase and the PRs in flight. Feel free to use or ignore my comments as you see fit. They are not meant to block your PR in any way.

File: internal/bird/monitor.go
Comment (arp-est) [reply internal/bird/monitor.go]: yeah, thanks good catch

File: internal/bird/monitor.go (hunk 7 above)
Comment (arp-est) [reply internal/bird/monitor.go:122]: I havent thought about it, previously we had only a single birdc command that could fail, I can add it.

File: internal/bird/monitor.go (hunk 8 above)
Comment (arp-est) [reply internal/bird/monitor.go:175]: For point 1, and 3, Renaming it to the same as the command and adding the -v flag to have some return codes seems like a good idea, but then I would need to update the parseProtocolOutput function name too, for consistency.
How about we make a follow up issue about these two things?

For point 2: There are units tests that already have sample birdc output I don't feel the need of an extra comment.

**Review by ljkiraly — APPROVED** [review]
Looks good now

File: internal/bird/monitor.go (hunk 7 above)
Comment (arp-est) [reply internal/bird/monitor.go:122]: oh wait, it seems to already distinguish the only two possible queries
`
log.V(1).Info("birdc query failed", "error", err) // show protocols

log.V(1).Info("birdc bfd query failed", "error", err) // show bfd sessions
   `
So I would prefer to leave it as is in this pr, we can add it to the follow up issue mentioned below

File: cmd/router/cmd/run.go [hunk 9]
```diff
@@ -244,33 +245,41 @@ func monitorConnectivity(
...
+			applyBfdState(status.Protocols, status.BfdSessions, routers, log)
+
 			count := protocolsUp(status.Protocols)
 
 			// Log when protocol up count changes
 			if firstUpdate || count != lastCount {
-				if status.HasConnectivity {
-					log.Info("Gateway connectivity established", "status", status.StatusString())
+				if count == 0 {
+					log.Info("Gateway connectivity ALL DOWN", "status", status.StatusString())
```
Comment (zolug) [inline cmd/router/cmd/run.go]: Wording is weird. Connectivity can be interpreted (and was in case of Meridio v1) as sg above Gateway Connections.
"All Gateway Connections Down", "All Gateway Connection Up", "Some Gateway Connection Down" or similar at this point would make more sense.

Btw, what happens here during monitoring if there are no GatewayRouters by config (i.e. len(routers)==0)? Probably a dedicated log would be worth for that case. Does it make sense to call applyBfdState() if len(routers) is zero?

File: docs/controllers/router.md [hunk 10]
```diff
@@ -124,13 +124,14 @@ Blackhole routes (`0.0.0.0/0` and `::/0`) in table 4097 prevent traffic leaking
...
+### BGP and BFD Monitoring
 
-A separate goroutine polls `birdc show protocols all "NBR-*"` at 1-second intervals:
+A separate goroutine polls `birdc show protocols all "NBR-*"` and `birdc show bfd sessions` at 1-second intervals:
 
 - Parses protocol state (up/down/start/idle) and BGP session info (Established)
-- Tracks connectivity: at least one `NBR-*` protocol in state `up` with info `Established`
-- Currently log-only; status changes logged when protocol up-count changes
+- For static protocols with BFD configured, applies BFD session state: if the BFD session is Up, Info is set to `Established`; if BFD is down or not found, Info is set to `BFD Down`
+- Tracks connectivity: at least one `NBR-*` protocol with Info containing `Established`
```
Comment (zolug) [inline docs/controllers/router.md]: Is this what gets exposed via the readiness gates? If so, would be nice to explain connectivity is handled separately for IPv4 and IPv6

File: internal/bird/monitor.go (hunk 7 above)
Comment (rubasov) [reply internal/bird/monitor.go:122]: Okay.

File: internal/bird/monitor.go (hunk 8 above)
Comment (rubasov) [reply internal/bird/monitor.go:175]: I'm completely okay with a followup. After commenting here I also found the examples in the tests, thanks!

File: cmd/router/cmd/run.go (hunk 9 above)
Comment (arp-est) [reply cmd/router/cmd/run.go]: I don't like the Some Gateway Connection Down message, also it would be nice for them to be uniform so that its easier to grep, and I couldn't come up with any neat name for It otherwise.

I think we talked about this too much, how about just print out one line with protocol count/total, and just get it done, and just do the discussion in the follow up.

File: docs/controllers/router.md (hunk 10 above)
Comment (arp-est) [reply docs/controllers/router.md]: discussed it, updated it to use your suggestion
